v2/cmd/controller/app/setup.go

```diff
@@ -44,6 +50,40 @@ import (
 	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/conditions"
 )
 
+func SetupPreUpgradeCheck(ctx context.Context) error {
+	cfg, err := clientconfig.GetConfig()
+	if err != nil {
+		return errors.Wrap(err, "unable to get client config")
+	}
+
+	apiExtClient, err := apiextensionsclient.NewForConfig(cfg)
+	if err != nil {
+		return errors.Wrap(err, "unable to create kubernetes client")
+	}
+
+	// Had to list CRDs this way and not with crdManager, since we did not have "serviceoperator.azure.com/version" labels in earlier versions.
+	list, err := apiExtClient.CustomResourceDefinitions().List(ctx, v1.ListOptions{})
+	if err != nil {
+		return errors.Wrap(err, "failed to list CRDs")
+	}
+
+	crdRegexp := regexp.MustCompile(`.*\.azure\.com`)
+	var errs []error
+	for _, crd := range list.Items {
+		crd := crd
+		if !crdRegexp.MatchString(crd.Name) {
+			continue
+		}
+
+		if policy, ok := crd.Annotations["helm.sh/resource-policy"]; !ok || policy != "keep" {
+			err = errors.New(fmt.Sprintf("looks like CRD '%s' does not contain helm keep policy. Make sure the upgrade is from beta.5", crd.Name))
```

Comment:
Suggested wording change. #nitpick

```suggestion
			err = errors.New(fmt.Sprintf("CRD '%s' does not have annotation for helm keep policy. Make sure the upgrade is from beta.5", crd.Name))
```